[FEATURE][ADMIN] Permettre la création d'organisations avec un identifiant externe déjà présent en base de données (PIX-9785)

igorissen opened the pull request:
## :unicorn: Problème

Il y a un besoin de créer plusieurs organisations avec le même identifiant externe, car il y a le besoin parfois de créer N orgas avec le même externalid (=SIRET). Mais pour le moment ce n'est pas possible avec Pix Admin OGA. Il existe une règle métier qui interdit la création de plusieurs organisations avec le même identifiant externe. Cette règle métier n'est validée que dans le usecase et pas comme contrainte en base de données.

Les agents qui doivent faire cette création malgré tout, doivent faire plusieurs manipulation pour y arriver :

1. ajouter une lettre ou un chiffre en plus pour rendre l'organisation unique
2. retirer cette lettre ou ce chiffre via metabase après création de l'organisation

## :robot: Proposition

Retirer la contrainte métier dans le usecase.

## :rainbow: Remarques

⚠️  Il faudrait vérifier que toutes les fonctionnalités dépendantes de l'identifiant externe (externalId) ne soient pas impactées par cette modification (ex: recherche avec un identifiant externe retournera potentiellement un tableau et plus un élément)

## :100: Pour tester

1. Se connecter sur **Pix Admin** (https://admin-pr8264.review.pix.fr/) avec un compte ayant le rôle **SUPER_ADMIN**
2. Cliquer sur l'onglet **Administration** (avant dernier élément du menu)
3. Cliquer sur le bouton **Importer un fichier CSV** du bloc **Création en masse d'organisations**
4. Importer un CSV contenant une organisation ayant le même identifiant externe (externalId) qu'une organisation existante en base de données. _Ne pas oublier d'ajouter l'id du compte sur lequel vous êtes connecté en tant que créateur de l'organisation (colonne createdBy)_.
5. Constater l'affichage d'une notification de succès
6. Effectuer l'étape 3 à 5 autant de fois que vous le souhaitez
7. Cliquer sur l'onglet **Organisations**
8. Filtrer la liste par **externalId** avec votre identifiant externe
9. Constater que les organisations que vous avez créé sont bien présentes

Voici un fichier CSV pour vos tests

```
type,externalId,name,provinceCode,credit,emailInvitations,emailForSCOActivation,organizationInvitationRole,locale,tags,createdBy,documentationUrl,targetProfiles,isManagingStudents,identityProviderForCampaigns,DPOFirstName,DPOLastName,DPOEmail
PRO,EXT_ID_DUP,PRO Orga 1,,,,,ADMIN,fr-fr,PUBLIC,90000,,,,,,,
PRO,EXT_ID_DUP,PRO Orga 2,,,,,ADMIN,fr-fr,PUBLIC,90000,,,,,,,
SCO,EXT_ID_DUP,SCO Orga 1,,,,,ADMIN,fr-fr,PUBLIC,90000,,,,,,,
SCO-1D,EXT_ID_DUP,SCO-1D Orga 1,,,,,ADMIN,fr-fr,PUBLIC,90000,,,,,,,
SCO-1D,EXT_ID_DUP,SCO-1D Orga 2,,,,,ADMIN,fr-fr,PUBLIC,90000,,,,,,,
```

igorissen committed:
refactor(api): remove useless usage of hasAtLeastOneAccessOf method

from securityPreHandlers module

igorissen committed:
docs(api): add missing "admin" tag

igorissen committed:
refactor(api): reorder properties and imports

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr8264.review.pix.fr)
  * [App (.org)](https://app-pr8264.review.pix.org)
  * [Orga (.fr)](https://orga-pr8264.review.pix.fr)
  * [Orga (.org)](https://orga-pr8264.review.pix.org)
  * [Certif (.fr)](https://certif-pr8264.review.pix.fr)
  * [Certif (.org)](https://certif-pr8264.review.pix.org)
  * [1D](https://1d-pr8264.review.pix.fr)
  * [Admin](https://admin-pr8264.review.pix.fr)
  * [API](https://api-pr8264.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr8264.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr8264/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr8264/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr8264/environment)

lego-technix commented on api/lib/application/organizations/index.js:
```diff
@@ -44,11 +44,7 @@ const register = async function (server) {
-                request,
-                h,
-              ),
+            method: securityPreHandlers.checkAdminMemberHasRoleSuperAdmin,
```
Modifier le message de ce commit dans le sens de la suggestion ci-après de manière à ne pas laisser penser qu'une vérification de sécurité est supprimée : 
`refactor(api): remove usage of hasAtLeastOneAccessOf method`
→ `refactor(api): remove useless usage of hasAtLeastOneAccessOf`

lego-technix commented on api/tests/unit/domain/usecases/create-organizations-with-tags-and-target-profiles_test.js:
```diff
@@ -484,31 +482,4 @@ describe('Unit | UseCase | create-organizations-with-tags-and-target-profiles',
     // then
     expect(organizationInvitationService.createProOrganizationInvitation).to.not.have.been.called;
   });
```
Il me semble qu'ajouter un test démontrant que créer des organisations avec les mêmes `externalId` est valide serait une bonne chose.

lego-technix commented on api/tests/unit/domain/usecases/create-organizations-with-tags-and-target-profiles_test.js:
```diff
@@ -484,31 +482,4 @@ describe('Unit | UseCase | create-organizations-with-tags-and-target-profiles',
     // then
     expect(organizationInvitationService.createProOrganizationInvitation).to.not.have.been.called;
   });
```
Et suivant la même idée, j'ai été gêné dans mon interprétation par le test `validation error cases - when there is more than one occurrence of the same organization in data file` qui me fait penser qu'il n'est pas valide de créer des organisations avec les mêmes `externalId`. Je ne dis pas du tout que le code est fautif, mais se serait une bonne chose si on pouvait reformuler le titre/modifier le code de ce test pour que son intention soit plus claire en ce qui concerne les `externalId`.

igorissen replied:
J'ai retiré ce test.

igorissen committed:
test(api): remove some unit tests

the new business rule accepts use of the same externalId for multiple organizations

igorissen committed:
test(api): reorder and rename contexts and tests

igorissen committed:
test(api): remove some integration tests

the new business rule accepts use of the same externalId for multiple organizations

igorissen committed:
test(api): rename contexts and tests

igorissen committed:
feat(api): remove "externalId" checks

Business rule changed. Now we can create multiple organizations with the same "externalId"

igorissen committed:
refactor(api): extract code to private function

igorissen committed:
feat(api): return created organization

with the data used to create the row in the database

igorissen committed:
feat(api): pass organization csv data

instead of organization for admin mapped data

igorissen committed:
feat(api): modify usecase

in order to take into account the new business rule for the "externalId"

igorissen committed:
feat(api): monitor errors

igorissen committed:
test(api): add missing test

lego-technix commented on api/tests/integration/domain/usecases/create-organizations-with-tags-and-target-profiles_test.js:
```diff
@@ -37,65 +34,7 @@ describe('Integration | UseCases | create-organizations-with-tags-and-target-pro
-    });
-
+  describe('#errors', function () {
     context('when there is organizations data file is empty', function () {
```
En profiter pour mettre à jour ce contexte qui est difficilement lisible : `when there is organizations data file is empty`

lego-technix commented on api/tests/integration/domain/usecases/create-organizations-with-tags-and-target-profiles_test.js:
```diff
@@ -802,4 +802,65 @@ describe('Integration | UseCases | create-organizations-with-tags-and-target-pro
     });
   });
+
+  context('when multiple organizations have the same "externalId"', function () {
```
Serait-il possible de mettre ce test dans le même commit qui rend possible la création de multiples organisations avec le même `externalId` ?

igorissen replied:
Si je déplace ce commit, les tests sur le commit qui recevra ce dernier ne fonctionnera pas du au faite que le code doit changer pour que ce test passe.

igorissen committed:
feat(api): modify joi schema

the current Joi schema does not follow correctly some business rules

lego-technix commented on api/lib/domain/usecases/create-organizations-with-tags-and-target-profiles.js:
```diff
@@ -98,7 +98,7 @@ function _transformOrganizationsCsvData(organizationsCsvData) {
       },
       emailInvitations: organizationCsvData.emailInvitations,
-      tags: organizationCsvData.tags.split(SEPARATOR),
+      tags: isEmpty(organizationCsvData.tags) ? [] : organizationCsvData.tags.split(SEPARATOR),
```
Pourrait-on dans ce commit `feat(api): modify joi schema ` expliquer dans le message du commit le pourquoi il y a cette modification du joi schema ? On s'en doute, mais ça permet d'être explicite. Et aussi ajouter une note à ce sujet dans la description de la PR :pray:

lego-technix commented on api/tests/unit/domain/usecases/create-organizations-with-tags-and-target-profiles_test.js:
```diff
@@ -1,9 +1,4 @@
-  OrganizationAlreadyExistError,
-  OrganizationTagNotFound,
-} from '../../../../lib/domain/errors.js';
+import { ObjectValidationError, OrganizationTagNotFound } from '../../../../lib/domain/errors.js';
```
Dans ce commit `test(api): remove some unit tests` il aurait été bon d'expliquer pourquoi ces tests sont enlevés : parce qu'ils vont à l'encontre de ce qu'on veut faire avec la PR, ou autre :pray:

lego-technix commented on api/tests/integration/domain/usecases/create-organizations-with-tags-and-target-profiles_test.js:
```diff
@@ -37,65 +35,7 @@ describe('Integration | UseCases | create-organizations-with-tags-and-target-pro
-    });
-
-    context('when there is organizations data file is empty', function () {
+    context('when provided CSV file is empty', function () {
```
Dans ce commit `test(api): remove some integration tests ` il aurait été bon d'expliquer pourquoi ces tests sont enlevés : parce qu'ils vont à l'encontre de ce qu'on veut faire avec la PR, ou autre 🙏

lego-technix reviewed (approved):
✅ Lu et testé fonctionnellement avec succès avec Firefox

EmmanuelleBonnemay reviewed (approved):
Lu et testé fonctionnellement avec succès avec Firefox.
